cub/test/catch2_test_device_reduce_nondeterministic.cu

```diff
@@ -0,0 +1,298 @@
+
+  constexpr auto min_op = cuda::minimum<type>{};
+  constexpr auto init   = cuda::std::numeric_limits<type>::max();
+
+  c2h::device_vector<type> d_input(num_items);
+  c2h::gen(C2H_SEED(2), d_input, static_cast<type>(-1000.0), static_cast<type>(1000.0));
+
+  const int num_indices = GENERATE_COPY(take(2, random(1 << 10, num_items)));
+
+  CAPTURE(num_indices, num_items);
+
+  // generate random indices to scatter NaNs
+  c2h::device_vector<int> indices(num_indices);
+  for (int i = 0; i < 2; ++i)
+  {
+    const type nan_val = i == 0 ? limits_t::signaling_NaN() : limits_t::quiet_NaN();
+
+    c2h::gen(C2H_SEED(2), indices, 0, num_items - 1);
+    auto begin = thrust::make_constant_iterator(nan_val);
+    auto end   = begin + num_indices;
+
+    // sprinkle some NaNs randomly throughout the input
+    thrust::scatter(c2h::device_policy, begin, end, indices.cbegin(), d_input.begin());
+  }
+
+  c2h::device_vector<type> d_output_p1(1);
+  c2h::device_vector<type> d_output_p2(1);
+
+  auto env1 = cuda::std::execution::env{
+    cuda::execution::require(cuda::execution::determinism::not_guaranteed), cuda::execution::__tune(hub_t<1, 128>{})};
+
+  auto env2 = cuda::std::execution::env{
+    cuda::execution::require(cuda::execution::determinism::not_guaranteed), cuda::execution::__tune(hub_t<2, 256>{})};
+
+  cub::DeviceReduce::Reduce(d_input.begin(), d_output_p1.begin(), num_items, min_op, init, env1);
+  cub::DeviceReduce::Reduce(d_input.begin(), d_output_p2.begin(), num_items, min_op, init, env2);
+
+  REQUIRE_EQ_WITH_NAN_MATCHING(d_output_p1, d_output_p2);
+}
+
+C2H_TEST("Nondeterministic Device reduce works with float and double on gpu with different iterators",
+         "[reduce][nondeterministic]",
+         float_type_list)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = 1 << 10;
+  const auto env      = cuda::execution::require(cuda::execution::determinism::not_guaranteed);
+
+  SECTION("device_vector iterators")
+  {
+    c2h::device_vector<type> d_input(num_items);
+    c2h::gen(C2H_SEED(2), d_input, static_cast<type>(-1000.0), static_cast<type>(1000.0));
+
+    c2h::device_vector<type> d_output(1);
+
+    cub::DeviceReduce::Reduce(d_input.begin(), d_output.begin(), num_items, cuda::std::plus<type>{}, type{}, env);
+
+    c2h::host_vector<type> h_input = d_input;
+
+    c2h::host_vector<type> h_expected(1);
+    h_expected[0] = std::accumulate(h_input.begin(), h_input.end(), type{}, ::cuda::std::plus<type>());
+    c2h::host_vector<type> h_output = d_output;
+
+    REQUIRE_APPROX_EQ_EPSILON(h_expected, h_output, type{0.01});
+  }
+
+  SECTION("constant iterator")
+  {
+    thrust::constant_iterator<type> input(1.0f);
+    c2h::device_vector<type> d_output(1);
+
+    cub::DeviceReduce::Reduce(input, d_output.begin(), num_items, cuda::std::plus<type>{}, type{}, env);
+
+    c2h::host_vector<type> h_expected(1);
+    h_expected[0] = std::accumulate(input, input + num_items, type{}, ::cuda::std::plus<type>());
```

Comment:
Suggestion: since we are reducing `num_items` times `1.0f`, I think we do not need to reduce on the host, but expect the output sum to be `num_items`. This should make the test faster.